docs(ROCm systems profiler): Creating Systems profiler overview doc (#9862)
ROCm/rocm-systems: projects/rocprofiler-systems/docs/quick-reference/rocprofiler-systems-at-a-glance.rst

.. meta::
   :description: A high-level overview of ROCm Systems Profiler architecture, instrumentation modes, and command-line tools
   :keywords: ROCm Systems Profiler overview, rocprofiler-systems, rocprof-sys, instrumentation, sampling, presets, rocpd, perfetto, MPI, OpenMP, GPU metrics

.. _rocprofiler-systems-at-a-glance:

=================================
ROCm Systems Profiler at a glance

> =================================

Reviewer: ```suggestion
******************************************************
ROCm Systems Profiler at a glance
******************************************************
```

ROCm Systems Profiler (``rocprofiler-systems``) captures the entire application stack in a single, unified timeline: CPU function calls, GPU kernel dispatches, MPI collectives, OpenMP regions, memory allocations, OS scheduling events, and GPU telemetry such as temperature, power, utilization, and interconnect bandwidth.

> This system-wide view matters most when the GPU is only one part of a larger workload. In a distributed training job, for example, the GPU can appear idle not because its kernels are slow, but because a data loader is starved, an MPI collective is blocking, or Python's GIL is stalling the dispatch queue. Diagnosing these cross-domain interactions requires a tool that sees the full picture, not just the GPU in isolation.

Reviewer 2: Can we instead say:

```suggestion
This system-wide view is especially useful when end-to-end performance depends on interactions across CPU, GPU, communication, and runtime - not on any single layer in isolation. In a distributed training job, for example, the GPU can appear idle not because its kernels are slow, but because a data loader is starved, an MPI collective is blocking, or Python's GIL is stalling the dispatch queue. Diagnosing these cross-domain interactions requires a tool that sees the full picture, not just the GPU in isolation.
```

> This topic orients you to how ROCm Systems Profiler is put together and how to invoke it. For the full, categorized feature catalog and use cases, see :doc:`../conceptual/rocprof-sys-feature-set`; for a narrative introduction to output formats, see :doc:`../what-is-rocprof-sys`

Reviewer: ```suggestion
This topic orients you to how ROCm Systems Profiler is put together and how to invoke it. For the full, categorized feature catalog and use cases, see :doc:`../conceptual/rocprof-sys-feature-set`; for a narrative introduction to output formats, see :doc:`../what-is-rocprof-sys`.
```

Author: The sentence ends with a "?".

.. _glance-key-features:

Key features
==============

The feature set is spread across several how-to and conceptual pages; this table is a quick index into them.

> 

Reviewer 2: This section lists several different concepts under the features category,
Some are collection mechanism, others are tracing capabilities etc.
In rocprof-sys-feature-set.rst, there is a separation into different categories.

We can either call this section something else or create sub-groups under each category and call it - Capabilities.

     - :doc:`../how-to/openmp-profiling`
   * - CPU hardware counters
     - Reads IPC, cache misses, branch mispredictions, and other CPU PMU events via the Linux ``perf`` subsystem.
     - :ref:`rocprof-sys-feature-cpu-metrics`
   * - GPU metrics
     - Samples GPU temperature, power, utilization, clocks, memory, and XGMI/PCIe bandwidth via ``amd-smi``; also collects GPU hardware counters via ``--gpu-events``.
     - :ref:`rocprof-sys-feature-gpu-metrics`, :doc:`../how-to/xgmi-pcie-sdma-sampling`, :doc:`../how-to/vcn-jpeg-sampling`
   * - Causal profiling
     - Estimates the end-to-end speedup from optimizing a given function or line by selectively slowing other code regions.
     - :doc:`../how-to/performing-causal-profiling`
   * - Python hooks
     - Instruments Python interpreter call frames for mixed Python/C++/HIP workload analysis.
     - :doc:`../how-to/profiling-python-scripts`
   * - Process attachment
     - Attaches to an already-running process without restarting it, similar to the ``rocprofv3`` attach feature.
     - :doc:`../how-to/attaching-to-running-process`

.. _glance-architecture:

How it works
=============

> ROCm Systems Profiler collects CPU-side data using one of two instrumentation modes, then collects GPU tracing, CPU counters, GPU metrics, MPI, OpenMP, and Python data through a shared data collection layer:

Reviewer 2: This could lead to a misunderstanding.

```suggestion
ROCm Systems Profiler collects data through several mechanisms - binary instrumentation, statistical sampling, callback APIs, and symbol interception - and merges CPU, GPU, communication, and system metrics into shared trace/profile output.
```

.. image:: ../data/how_systems_profiler_works.png
   :width: 60%
   :align: center

For the full explanation of each collection mode, including overhead trade-offs and a worked instrumentation-vs-sampling example, see :doc:`../conceptual/data-collection-modes`.

.. _glance-modes-compared:

Instrumentation modes compared
================================

> Binary instrumentation and call-stack sampling can be used independently or together, trading overhead against completeness:

Reviewer 2: Sampling is not an instrumentation mode.
We can call this section either “Binary instrumentation vs statistical sampling” or something more appropriate.

.. list-table::
   :header-rows: 1

   * - Mode
     - Overhead
     - Completeness
     - Best for
   * - Binary instrumentation
     - High
     - Complete (every call recorded)
     - Function-level profiling when overhead is acceptable
   * - Call-stack sampling
     - Low
     - Statistical (sampled)
     - Production profiling, long-running jobs, MPI workloads
   * - Combined (both enabled)

>      - Medium

Reviewer: ```suggestion
     - High
```

Reviewer: If Binary Instrumentation is high, how does adding Call-stack sampling make it less high?

     - Best of both
     - When deterministic detail and low-overhead statistical coverage are both needed

.. _glance-cli-tools:

Command-line tools
====================

> ROCm Systems Profiler ships as a set of standalone executables, one per collection mode:

Reviewer 2: All executable are not different collection modes.

```suggestion
ROCm Systems Profiler ships as a set of standalone executables, each oriented toward a different profiling workflow:
```

.. list-table::
   :header-rows: 1
   :widths: 15 35 25 25

   * - Tool
     - Purpose
     - Example
     - See also
   * - ``rocprof-sys-avail``
     - Lists configuration settings, hardware counters, and data-collection components
     - ``rocprof-sys-avail``
     - :doc:`../how-to/general-tips-using-rocprof-sys`
   * - ``rocprof-sys-instrument``
     - Performs dynamic binary instrumentation, including binary rewriting
     - ``rocprof-sys-instrument -o ./app.inst -- ./app``
     - :doc:`../how-to/instrumenting-rewriting-binary-application`
   * - ``rocprof-sys-run``
     - Launches a binary-rewritten executable, or profiles an application directly using preset/domain flags
     - ``rocprof-sys-run -- ./app.inst``
     - :doc:`../how-to/instrumenting-rewriting-binary-application`, :ref:`using-preset-profiles-quick-start`
   * - ``rocprof-sys-sample``
     - Performs call-stack sampling without instrumentation
     - ``rocprof-sys-sample -f 1000 -- ./app``
     - :doc:`../how-to/sampling-call-stack`
   * - ``rocprof-sys-attach``
     - Attaches to an already-running process
     - ``rocprof-sys-attach -p $(pidof my_app)``
     - :doc:`../how-to/attaching-to-running-process`

> 

Reviewer 2: `rocprof-sys-causal` is missing.

MPI workloads are typically run with ``rocprof-sys-sample`` under ``mpirun``:

.. code-block:: shell

>    mpirun -n 4 rocprof-sys-sample --preset=trace-hpc -- ./mpi_app

Reviewer 2: This `mpirun` example seems to be a misfit in this section.

Author: Removed it from here. This content is already covered in [using-preset-profiles.rst].

.. _glance-presets:

> Preset profiles

Reviewer 2: **“Preset profiles”** and **“Available presets”** need not be separate sections.

Author: Merged them into one.

.. meta::
  :description: A high-level overview of ROCm Systems Profiler architecture, instrumentation modes, and command-line tools
  :keywords: ROCm Systems Profiler overview, rocprofiler-systems, rocprof-sys, instrumentation, sampling, presets, rocpd, perfetto, MPI, OpenMP, GPU metrics

.. _rocprofiler-systems-at-a-glance:

=================================

> ROCm Systems Profiler at a glance

Reviewer 3: I feel like this "ROCm Systems Profiler at a glance" and the existing "What is?" page serve the same purpose... I think it would be better to combine the two? @SwRaw

Author: @peterjunpark What-is-systems-profiler page doesn't need to be so detailed. It is for users who just want to know what it is about, while the "At-a-glance" page summarizes all the important features and a glimpse of the overall architecture (which might be too much for someone who simply wants to know what the tool is about). This doc also serves as a map for users trying to learn everything about the component at once, and from there they can delve into the details of respective features via links to the various pages.
